Update Mosviz parser doc and fix NIRSpec auto-load

pllim opened the pull request:
<!-- This comments are hidden when you submit the pull request,
so you do not need to remove them! -->

<!-- Please be sure to check out our code of conduct,
https://github.com/spacetelescope/jdaviz/blob/main/CODE_OF_CONDUCT.md . -->

### Description
<!-- Provide a general description of what your pull request does.
Complete the following sentence and add relevant details as you see fit. -->

<!-- In addition please ensure that the pull request title is descriptive
and allows maintainers to infer the applicable viz component(s). -->

This pull request is to update Mosviz doc on importing data, and also enable NIRSpec auto loader to take a single image.

And then the scope expanded and now it also fixes #2144

### Change log entry

- [x] Is a change log needed? If yes, is it added to `CHANGES.rst`? If you want to avoid merge conflicts,
  list the proposed change log here for review and add to `CHANGES.rst` before merge. If no, maintainer
  should add a `no-changelog-entry-needed` label.

### Checklist for package maintainer(s)
<!-- This section is to be filled by package maintainer(s) who will
review this pull request. -->

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [x] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [x] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)? Not sure how to add tests that don't require downloading a bunch of remote data files. Can we not?
- [x] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Did the CI pass? If not, are the failures related?
- [x] Is a milestone set? Set this to bugfix milestone if this is a bug fix and needs to be released ASAP; otherwise, set this to the next major release milestone.
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)? [🐱](https://jira.stsci.edu/browse/JDAT-3041) [🐱](https://jira.stsci.edu/browse/JDAT-3234) [🐱](https://jira.stsci.edu/browse/JDAT-3287)

camipacifici commented:
Thank you very much. This fixes my original problem.
I see three additional problems here:
- the slit overlay is likely wrong (and this is probably not a jdaviz problem...lets just ignore this for now)
- the table is returning `Unspecified` for all the sources under the RA and Dec columns
- the position where the image viewer is zooming to is different from the position given by the SRCRA and SRCDEC keywords. I am not 100% positive that these keywords are the right ones, but maybe this can be double checked with the NIRISS parser that I am sure is correct.

camipacifici commented:
Would be nice to fix the image loader for this one too:
https://github.com/spacetelescope/jdaviz/issues/2144

pllim commented:
> the table is returning Unspecified for all the sources under the RA and Dec columns

For some reason, it is trying to look for RA and Dec on the image, not spectra. Not sure why. I added a fix.

> the position where the image viewer is zooming to is different from the position given by the SRCRA and SRCDEC keywords

Hard for me to test using the cutout off the 2GB image you provided. If this is still a problem, please be more specific on what the correct answer should be. In NIRISS, I see that the parser grab RA and Dec off a special ECSV catalog that NIRSpec does not have, so I cannot just use the NIRISS logic. And are you sure this is not the same problem as S_REGION? These are all populated by the pipeline.

> Would be nice to fix the image loader

I also added a fix for this. Please test your use case in #2144 with this PR too.

pllim commented on jdaviz/configs/mosviz/helper.py:
```diff
@@ -659,10 +668,13 @@ def load_images(self, data_obj, data_labels=None, share_image=0):
                           data_labels=data_labels, share_image=share_image)
-        self._add_redshift_column()
+        if add_redshift_column:
+            self._add_redshift_column()
```
Note: Multiple calls to `self._add_redshift_column()` in `load_data` was what causing all the warnings Cami was seeing, so I have to disable this call in sub-calls by default to get rid of the warnings.

camipacifici replied:
I confirm I do not see the warnings anymore.

camipacifici commented on jdaviz/configs/mosviz/plugins/parsers.py:
```diff
@@ -490,8 +509,20 @@ def mos_meta_parser(app, data_obj=None, ids=None):
+        # Grab target sky coordinates from 1D spectrum, if possible.
+        # This has to happen after Filter/Grating because columns are insertion-ordered.
+        if has_1d:
+            ra = query_metadata_by_component(app, "SRCRA", "1D Spectra", False)
```
I took a deeper look at the header and it seems like SRCRA and SRCDEC are the right keywords.

camipacifici commented:
The table is nicely populated, but if I try to load the image I still get the same traceback that ends with
ValueError: The dimensions of component 2D Spectra are incompatible with the dimensions of this data: (50,) vs (1,)

This is the way I am calling it:
`mosviz = Mosviz()`
`mosviz.load_data(spectra_1d="../data/mastDownload/jones_calx1d_files/single/jw02136001001_03137_00001_nrs1_x1d.fits",
                 spectra_2d="../data/mastDownload/jones_s2d_level2/jw02136001001_03137_00001_nrs1_s2d.fits",
                 images="/Users/cpacifici/Dropbox/SINCRONIZZARE/Documents/WORK/CANDELS/hlsp_candels_hst_wfc3_egs-tot-60mas_f160w_v1.0_drz.fits",
                 instrument="nirspec")`
`mosviz.show()`

pllim commented on jdaviz/configs/mosviz/tests/test_data_loading.py:
```diff
@@ -212,7 +209,7 @@ def test_load_single_image_multi_spec(mosviz_helper, mos_image, spectrum1d, mos_
                                         {'event': 'mousemove', 'domain': {'x': 10, 'y': 100}})
     assert label_mouseover.as_text() == ('Pixel x=00010.0 y=00100.0 Value +8.12986e-01', '', '')
-    assert label_mouseover.icon == 'b'
+    assert label_mouseover.icon == 'c'
```
Note: I changed the loading order in load_data to fix a bug, hence this change here.

pllim commented:
> The image viewer is not zooming on SRCRA and SRCDEC

I see the same thing but I dunno what the real problem is. The image emits a lot of WCS warnings when you load it. I think its WCS might not be entirely correct.

Also I don't know if SRCRA and SRCDEC suffer from the same problem as S_REGION since they are all populated by the same pipeline; have you heard back from NIRSpec/pipeline about this?

If you have another image that is more well-behaved and not 2GB, we can try to test this but on that image to be sure.

pllim commented:
> what was the way that worked for you

I was loading using directory, having moved the image under some sub-dir that it looks for. I did not try to load all 3 components separately and when I did, I could reproduce your problem (and then fix it).

pllim commented:
Hmm I just found the code that zoom on shared image. I forgot this is not Imviz, so maybe that is a real bug, lemme see.

camipacifici commented:
I cannot find a decent image the covers the appropriate region sorry.
If you think there is a bug, I will test any fix you might have.
Thank you!

pllim commented:
I think I did find a bug, where I missed that I have to set a private attribute in one of the logic route. 🤪 

There are several things going on at once with your slit and your target coordinates:

* The image WCS is not great. I had to manually remove the offending SIP coefficients (header keywords that starts with `A_` or `B_`) in the image file. Otherwise, the sky/pix transformation is suspicious and full of ~terror~ warnings.

```python
pf = fits.open(filename, mode="update")
for k in list(pf[0].header):
    if k.startswith('A_') or k.startswith('B_'):
        pf[0].header.remove(k)
pf.close()
```

* The image WCS is defined in FK5 but pretty sure JWST pipeline only knows ICRS, so I dunno, your pix/sky/pix/sky conversions might get confused in Jdaviz. Hopefully it is close enough that your target will still be in view. 🤷 

* As you know, the slit positions are wrong and unrelated to this PR, so I cannot rely on that to see how well centering is doing, so I had to make Mosviz image viewer inherit things from Astrowidgets (I think I feel Kyle's eyebrows twitching) so I can do the following sanity check:

```python
imview = mosviz.app.get_viewer(mosviz._default_image_viewer_reference_name)
imview._mark_targets()
```

So, with my new commits to this PR, you will be able to see that NIRSpec can now zoom to the correct target on table click (it is obvious after you use that hidden method to add markers). And that the slits are off. Since your cutout that I am using has the image data way off to a distance, I cannot see if the RA and Dec is actually marking any real targets, but at least it zooms to what the table reports.

I have a hard time finding good data for different instruments to this PR with. And there are way too many combos to test for each. And for some reason, Mosviz loader is very slow on my machine. So, I think you and other devs need to review this carefully even if it works for your particular NIRSpec case.

camipacifici commented:
It works!! Thank you for the snippet to fix the image too.
<img width="749" alt="Screen Shot 2023-04-13 at 10 03 01 PM" src="https://user-images.githubusercontent.com/19998720/231923002-29321a18-5c26-47bb-a877-2c4090e5d4bb.png">

camipacifici commented:
I will test with other datasets I can find and let you know how it goes.

camipacifici commented:
You are right!! My mistake, the image is there and the zoom is doing what it is supposed to, i.e. zooming using the number of pixels in the cross dispersion direction in the 2D image. That is apparently too small, but it not a jdaviz problem. Also the fact that the RA/Dec are not in the header is not a jdaviz problem.

For the record, I did not have to put the image under an `images` directory. I just included it in the folder, like for NIRISS, so docs are also correct.

Sorry for the false alarm! This looks good to me!

pllim commented:
Failures are unrelated.

rosteen commented:
I get a test failure that appears related when I run them locally, from `test_nirspec_parser[nirspec]`:

```
>       assert len(mosviz_helper.app.data_collection) == 16
E       AssertionError: assert 11 == 16
E        +  where 11 = len(DataCollection (11 data sets)\n	  0: MOS Table\n	  1: 1D Spectrum 0\n	  2: 1D Spectrum 1\n	  3: 1D Spectrum 2\n	  4: 1D Spe...	  5: 1D Spectrum 4\n	  6: 2D Spectrum 0\n	  7: 2D Spectrum 1\n	  8: 2D Spectrum 2\n	  9: 2D Spectrum 3\n	 10: 2D Spectrum 4)
E        +    where DataCollection (11 data sets)\n	  0: MOS Table\n	  1: 1D Spectrum 0\n	  2: 1D Spectrum 1\n	  3: 1D Spectrum 2\n	  4: 1D Spe...	  5: 1D Spectrum 4\n	  6: 2D Spectrum 0\n	  7: 2D Spectrum 1\n	  8: 2D Spectrum 2\n	  9: 2D Spectrum 3\n	 10: 2D Spectrum 4 = Application(config='mosviz', events=['call_viewer_method', 'close_snackbar_message', 'data_item_remove', 'data_item_un...erflow hidden will prevent the scrollbar from\n   * appearing */\n  overflow: hidden;\n}\n</style>\n'), vdocs='latest').data_collection
E        +      where Application(config='mosviz', events=['call_viewer_method', 'close_snackbar_message', 'data_item_remove', 'data_item_un...erflow hidden will prevent the scrollbar from\n   * appearing */\n  overflow: hidden;\n}\n</style>\n'), vdocs='latest') = <jdaviz.configs.mosviz.helper.Mosviz object at 0x13f810550>.app

configs/mosviz/tests/test_parsers.py:38: AssertionError
```

I assume that you weren't seeing this locally, any idea what might be going wrong?

pllim commented:
@rosteen , I seen that before. I think some data was not loaded properly. I thought I fixed that. Are you testing an older version of my branch? Did you try clearing the download cache?

rosteen commented on jdaviz/configs/mosviz/plugins/viewers.py:
```diff
@@ -28,23 +29,33 @@ class MosvizImageView(JdavizViewerMixin, BqplotImageView):
-    def set_plot_axes(self):
-        self.figure.axes[1].tick_format = None
-        self.figure.axes[0].tick_format = None
+    def _mark_targets(self):
```
Where does this get used, anyway? This is the only instance of the string "_mark_targets" that I see in the repository.

pllim replied:
I needed it to debug. I wanted to see where the RA and Dec from the table actually are. I feel like we could make it public (in one way or another) in the future but I don't want that headache right now.

It comes "for free" with the Astrowidgets inheritance, but if this bothers you, I can remove it.

rosteen replied:
Ah, I see. I don't think it needs to be removed then, but it would be nice to have a comment saying that it's there to help with debugging and isn't actually called internally.

duytnguyendtn replied:
I could see this being useful eventually, why not 🤷

pllim replied:
Comment added.

rosteen commented on jdaviz/configs/mosviz/plugins/viewers.py:
```diff
@@ -28,23 +29,33 @@ class MosvizImageView(JdavizViewerMixin, BqplotImageView):
 
-        # Make it so y axis label is not covering tick numbers.
-        self.figure.axes[1].label_offset = "-50"
+        from astropy.coordinates import SkyCoord
+        from astropy.table import QTable
```
Depends on the answer to the question above probably, but maybe move these to the top level? I generally dislike imports in functions.

pllim replied:
Does https://github.com/spacetelescope/jdaviz/pull/2146#discussion_r1170444678 help you making a decision on this?

rosteen replied:
If it's only used in debugging it can stay.

duytnguyendtn replied:
The comment doesn't quite give me an answer as to why these imports should be down here. Is there a reason?

pllim replied:
Because the imports are unnecessary otherwise. If we decide to delete this method one day, it gives a cleaner diff. If you really want the imports to be high up, I can do it too. Lemme know.

duytnguyendtn replied:
Understandable, but isn't this directly against PEP8? https://peps.python.org/pep-0008/#imports

> Imports are always put at the top of the file, just after any module comments and docstrings, and before module globals and constants.

Unless those particular imports take a particularly long time, I'd argue following PEP8. I'm personally comfortable with counting this as a value-add rather than a "small debugging thing that might be useful later"

pllim replied:
Imports moved up.

rosteen reviewed (approved):
I left a couple small questions and one small suggestion, but otherwise I think this looks good.

duytnguyendtn commented on jdaviz/configs/mosviz/helper.py:
```diff
@@ -288,31 +290,32 @@ def _row_click_message_handler(self, msg):
-            else:
-                try:
-                    center, height = self._zoom_to_slit_params(msg)
-                except IndexError:
-                    # If there's nothing in the spectrum2d viewer, we can't get slit info
-                    return
```
I don't see an equivalent to this catch. Was it a conscious decision to un-handle this exception?

pllim replied:
This is now caught within `_zoom_to_slit_params`. If somehow we still get an error here, it is a bug and we should just let the exception raise naturally.

duytnguyendtn commented on jdaviz/configs/mosviz/helper.py:
```diff
@@ -288,31 +290,32 @@ def _row_click_message_handler(self, msg):
+        if height is not None:
+            image_axis_ratio = (abs(imview.state.x_max - imview.state.x_min) /
+                                abs(imview.state.y_max - imview.state.y_min))
+            xh = image_axis_ratio * height
```
I generally have a disdain towards two-letter vars that aren't self explanatory. Any way we could add a meaningful name to `xh`?

pllim replied:
Changed.

duytnguyendtn reviewed (approved):
For whatever reason, I couldn't get to Cami's nirspec example, but the nircam example worked and zoomed well. Code looks good and much more logical. Thanks Pey-Lian! A few other comments

pllim commented:
@rosteen and @duytnguyendtn , is this okay to merge now? Thanks for the reviews!
